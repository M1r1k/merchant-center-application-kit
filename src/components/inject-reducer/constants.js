// eslint-disable-next-line import/prefer-default-export
export const ACTIVATE_PLUGIN = 'ACTIVATE_PLUGIN';
