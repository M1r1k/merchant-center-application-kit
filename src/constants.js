// eslint-disable-next-line import/prefer-default-export
export const USER_PROFILE_FORM_NAME =
  'packages-shared/app-shell/USER_PROFILE_FORM_NAME';
