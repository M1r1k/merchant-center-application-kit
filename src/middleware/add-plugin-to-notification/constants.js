export const __LOCAL = 'LOCAL';
export const __GLOBAL = 'GLOBAL';
