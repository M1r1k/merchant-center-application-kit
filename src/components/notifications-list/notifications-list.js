import PropTypes from 'prop-types';
import React from 'react';
import { removeNotification } from '@commercetools-local/notifications';
import { DOMAINS } from '@commercetools-local/constants';
import GenericNotification from '../notification-kinds/generic';
import ApiErrorNotification from '../notification-kinds/api-error';
import UnexpectedErrorNotification from '../notification-kinds/unexpected-error';
import styles from './notifications-list.mod.css';

function mapNotificationToComponent(notification) {
  switch (notification.domain) {
    case DOMAINS.PAGE:
      switch (notification.kind) {
        case 'error':
        case 'warning':
        case 'info':
        case 'success':
          return GenericNotification;
        case 'api-error':
          return ApiErrorNotification;
        case 'unexpected-error':
          return UnexpectedErrorNotification;
        default:
          return null;
      }
    case DOMAINS.GLOBAL:
      switch (notification.kind) {
        case 'error':
        case 'warning':
        case 'info':
        case 'success':
          return GenericNotification;
        case 'unexpected-error':
          return UnexpectedErrorNotification;
        default:
          return null;
      }
    case DOMAINS.SIDE:
      switch (notification.kind) {
        case 'info':
        case 'success':
        case 'warning':
        case 'error':
          return GenericNotification;
        default:
          return null;
      }
    default:
      return null;
  }
}

export class NotificationsList extends React.PureComponent {
  static displayName = 'NotificationsList';

  static propTypes = {
    domain: PropTypes.string.isRequired,
    notifications: PropTypes.arrayOf(PropTypes.object).isRequired,
    mapPluginNotificationToComponent: PropTypes.func,
    activePlugin: PropTypes.string,
    // Needed for global notifications at the moment.
    showUnexpectedErrorNotification: PropTypes.func,
  };

  static defaultProps = {
    mapPluginNotificationToComponent: null,
  };

  static contextTypes = {
    store: PropTypes.object.isRequired,
  };

  render() {
    return (
      <div className={styles[`container-${this.props.domain}`]}>
        {this.props.notifications.map(notification => {
          // check whether the current plugin provides a custom
          // notification for this type
          const PluginNotification =
            this.props.mapPluginNotificationToComponent &&
            this.props.mapPluginNotificationToComponent(notification);
          // fall back to app-wide notifications
          const Component =
            PluginNotification || mapNotificationToComponent(notification);
          if (!Component) {
            if (process.env.NODE_ENV !== 'production')
              // eslint-disable-next-line no-console
              console.error(
                `Saw unexpected notification kind "${notification.kind}".`,
                notification
              );
            return null;
          }
          const dismiss = () => {
            this.context.store.dispatch(removeNotification(notification.id));
          };
          return (
            <Component
              key={notification.id}
              notification={notification}
              dismiss={dismiss}
              // Needed for global notifications at the moment.
              showUnexpectedErrorNotification={
                this.props.showUnexpectedErrorNotification
              }
            />
          );
        })}
      </div>
    );
  }
}

export default NotificationsList;
