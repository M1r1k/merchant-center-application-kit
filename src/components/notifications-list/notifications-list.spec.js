import { shallow } from 'enzyme';
import React from 'react';
import { DOMAINS } from '@commercetools-local/constants';
import { NotificationsList, isNotificationVisible } from './notifications-list';

const createTestProps = props => ({
  domain: DOMAINS.PAGE,
  notifications: [],
  activePlugin: null,
  ...props,
});

describe('rendering', () => {
  it('outputs correct tree', () => {
    const props = createTestProps({
      notifications: [
        { id: 1, domain: DOMAINS.PAGE, kind: 'success', text: 'some-text' },
        { id: 2, domain: DOMAINS.PAGE, kind: 'info', text: 'some-other-text' },
        { id: 3, domain: DOMAINS.GLOBAL, kind: 'info', text: 'some-more-text' },
      ],
    });
    const wrapper = shallow(<NotificationsList {...props} />, {
      context: { store: { dispatch: jest.fn() } },
    });
    expect(wrapper).toMatchSnapshot();
  });
  it('displays notifications', () => {
    const props = createTestProps({
      notifications: [
        { id: 1, domain: DOMAINS.PAGE, kind: 'success', text: 'some-text' },
        { id: 2, domain: DOMAINS.PAGE, kind: 'info', text: 'some-other-text' },
      ],
    });
    const wrapper = shallow(<NotificationsList {...props} />, {
      context: { store: { dispatch: jest.fn() } },
    });
    expect(wrapper).toMatchSnapshot();
  });
});

describe('mapPluginNotificationToComponent', () => {
  it('can return a component that is preferred to internal kinds', () => {
    const FooNotification = () => <div />;
    const props = createTestProps({
      notifications: [
        { id: 1, domain: DOMAINS.PAGE, kind: 'success', text: 'some-text' },
        { id: 2, domain: DOMAINS.PAGE, kind: 'info', text: 'some-other-text' },
      ],
    });
    const wrapper = shallow(
      <NotificationsList
        {...props}
        mapPluginNotificationToComponent={notification =>
          notification.kind === 'success' ? FooNotification : null
        }
      />,
      { context: { store: { dispatch: jest.fn() } } }
    );
    expect(wrapper).toRender(FooNotification);
  });

  describe('side-notifications', () => {
    let props;
    let wrapper;

    beforeEach(() => {
      props = createTestProps({
        notifications: [
          {
            id: 1,
            domain: DOMAINS.SIDE,
            kind: 'success',
            text: 'some-success-text',
          },
          {
            id: 2,
            domain: DOMAINS.SIDE,
            kind: 'info',
            text: 'some-info-text',
          },
          {
            id: 3,
            domain: DOMAINS.SIDE,
            kind: 'warning',
            text: 'some-warning-text',
          },
          {
            id: 4,
            domain: DOMAINS.SIDE,
            kind: 'error',
            text: 'some-error-text',
          },
        ],
      });
      wrapper = shallow(<NotificationsList {...props} />, {
        context: { store: { dispatch: jest.fn() } },
      });
    });

    it('should map the message kind to the correct components', () => {
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.find('GenericNotification')).toHaveLength(4);
    });
  });
});

describe('isNotificationVisible', () => {
  describe('when no plugin is active', () => {
    it('should show global notifications', () => {
      expect(isNotificationVisible(null, null)).toBe(true);
    });
    it('should hide plugin notifications', () => {
      expect(isNotificationVisible(null, 'non-active-plugin')).toBe(false);
    });
  });
  describe('when a plugin is active', () => {
    it('should show global notifications', () => {
      expect(isNotificationVisible('active-plugin', null)).toBe(true);
    });
    it('should hide notifications from other plugins', () => {
      expect(isNotificationVisible('active-plugin', 'non-active-plugin')).toBe(
        false
      );
    });
    it('should show notifications from the active plugin', () => {
      expect(isNotificationVisible('active-plugin', 'active-plugin')).toBe(
        true
      );
    });
  });
});
