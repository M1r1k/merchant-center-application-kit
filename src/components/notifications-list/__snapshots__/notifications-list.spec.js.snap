// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapPluginNotificationToComponent side-notifications should map the message kind to the correct components 1`] = `
<div
  className="container-page"
>
  <GenericNotification
    dismiss={[Function]}
    key="1"
    notification={
      Object {
        "domain": "side",
        "id": 1,
        "kind": "success",
        "text": "some-success-text",
      }
    }
  />
  <GenericNotification
    dismiss={[Function]}
    key="2"
    notification={
      Object {
        "domain": "side",
        "id": 2,
        "kind": "info",
        "text": "some-info-text",
      }
    }
  />
  <GenericNotification
    dismiss={[Function]}
    key="3"
    notification={
      Object {
        "domain": "side",
        "id": 3,
        "kind": "warning",
        "text": "some-warning-text",
      }
    }
  />
  <GenericNotification
    dismiss={[Function]}
    key="4"
    notification={
      Object {
        "domain": "side",
        "id": 4,
        "kind": "error",
        "text": "some-error-text",
      }
    }
  />
</div>
`;

exports[`rendering displays notifications 1`] = `
<div
  className="container-page"
>
  <GenericNotification
    dismiss={[Function]}
    key="1"
    notification={
      Object {
        "domain": "page",
        "id": 1,
        "kind": "success",
        "text": "some-text",
      }
    }
  />
  <GenericNotification
    dismiss={[Function]}
    key="2"
    notification={
      Object {
        "domain": "page",
        "id": 2,
        "kind": "info",
        "text": "some-other-text",
      }
    }
  />
</div>
`;

exports[`rendering outputs correct tree 1`] = `
<div
  className="container-page"
>
  <GenericNotification
    dismiss={[Function]}
    key="1"
    notification={
      Object {
        "domain": "page",
        "id": 1,
        "kind": "success",
        "text": "some-text",
      }
    }
  />
  <GenericNotification
    dismiss={[Function]}
    key="2"
    notification={
      Object {
        "domain": "page",
        "id": 2,
        "kind": "info",
        "text": "some-other-text",
      }
    }
  />
  <GenericNotification
    dismiss={[Function]}
    key="3"
    notification={
      Object {
        "domain": "global",
        "id": 3,
        "kind": "info",
        "text": "some-more-text",
      }
    }
  />
</div>
`;
